File: KaraokeShow/Internalization/InternationalizationManager.vb
Imports System.Windows.Forms

Public Class InternationalizationManager
    Public Shared MB_GetLocalizationAPI As Func(Of String, String, String)
    Public Shared Function GetCurrentMusicBeeLanguage() As String
        'Initialize langDict that convert MusicBee setting to .Net Culture Text
        Dim langDict As New Dictionary(Of String, String)
        langDict("Language") = "en"
        langDict("语言") = "zh-Hans"
        langDict("言語") = "ja"
        'Get culture text
        Dim mbLang = MB_GetLocalizationAPI?.Invoke("Main.field.173", "Language")
        Dim cultureText = If(langDict.ContainsKey(mbLang), langDict(mbLang), "en")
        Return cultureText
    End Function
    Public Shared Sub EnableLanguage()
        Dim cultureText = InternationalizationManager.GetCurrentMusicBeeLanguage()
        Threading.Thread.CurrentThread.CurrentUICulture = New Globalization.CultureInfo(cultureText)
    End Sub
    Public Shared Sub ApplyResourceToWinForm(c As Control)
        Dim res = New EmbedResourceManager(c.GetType())
        For Each item As Control In c.Controls
            res.ApplyResources(item, item.Name)
        Next
        c.ResumeLayout(False)
        c.PerformLayout()
        res.ApplyResources(c, c.Name)
    End Sub
End Class

File: KaraokeShow/Internationalization/InternationalizationManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Globalization;
using System.Windows.Forms;
using System.Resources;

namespace MusicBeePlugin.Internationalization
{
    class InternationalizationManager
    {
        private static ResourceManager resMan = new EmbedResourceManager(typeof(Properties.Resources));
        public static string CultureText { get; set; } = "en";
        public static void SetCurrentLanguage(string mbMainField173Text)
        {
            //Initialize langDict that convert MusicBee setting to .Net Culture Text
            var langDict = new Dictionary<string, string>();
            langDict["Language"] = "en";
            langDict["语言"] = "zh-Hans";
            var cultureText = langDict.ContainsKey(mbMainField173Text) ? langDict[mbMainField173Text] : "en";
            CultureText = cultureText;
        }

        public static void EnableLanguage()
        {
            Thread.CurrentThread.CurrentUICulture = new CultureInfo(CultureText);
        }

        public static string GetResourceString(string stringName)
        {
            return resMan.GetString(stringName, new CultureInfo(CultureText));
        }

        public static void ApplyResourceToWinForm(Control c)
        {
            var res = new EmbedResourceManager(c.GetType());
            ApplyResourceToControl(c, res);
        }

        private static void ApplyResourceToControl(Control c, EmbedResourceManager res)
        {
            foreach (Control item in c.Controls)
            {
                ApplyResourceToControl(item, res);
            }
            res.ApplyResources(c, c.Name);
            c.ResumeLayout(false);
            c.PerformLayout();
        }

    }
}
